Javashim

xspeedcruiser opened the pull request:
Java chaincode support for Hyperledger Fabric
## Description

This PR is to extend fabric to be able to deploy and run chaincode written in Java.
## Motivation and Context

<!-- Why is this change required? What problem does it solve? -->

<!-- If it fixes an open issue, please link to the issue here. -->

Fabric project currently supports chaincode written in the Go language. This PR is to extend the capability of Fabric to run chaincode written in Java. Many organizations are already comfortable with the Java language and its tooling, have existing code bases written in Java, and have a large number of developers who are familiar with the language. Therefore, it would be useful to be able to write chaincodes in the Java language.
Fixes # 
## How Has This Been Tested?

<!-- If this PR does not contain a new test case, explain why. -->

<!-- Describe in detail how you tested your changes. -->

Simple unit test and Behave test written for a single peer scenario
## Checklist:

<!-- To check a box, and an 'x': [x] -->

<!-- To uncheck box, add a space: [ ] -->

<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff).
- [x] I have either added documentation to cover my changes or this change requires no new documentation.
- [x] I have either added unit tests to cover my changes or this change requires no new tests.
- [x] I have run [golint](https://github.com/golang/lint) and have fixed valid warnings in code I have added or modified. This tool generates false positives so you may choose to ignore some warnings. The goal is clean, consistent, and readable code.

The continuous integration build process will run [make checks](https://github.com/hyperledger/fabric/blob/master/Makefile#L22) to confirm that tests pass and that code quality meets minimum standards. You may optionally run this locally as PRs will not be accepted until they pass.

Signed-off-by: satheesh.ceg@gmail.com

Commit:
Prepares the peer to deploy Java-based chaincodes

Commit:
Prepares the peer to deploy Java-based chaincodes

Commit:
Prepares the peer to deploy Java-based chaincodes

Commit:
Completes provisioning for java chaincode

Commit:
Completes provisioning for java chaincode

Commit:
Completes provisioning for java chaincode

Commit:
Adds Java shim to main fabric project. Location is likely temporary.

Commit:
Adds Java shim to main fabric project. Location is likely temporary.

Commit:
Adds Java shim to main fabric project. Location is likely temporary.

Commit:
Removed grpc files that should be generated rather than stored

Commit:
Removed grpc files that should be generated rather than stored

Commit:
Removed grpc files that should be generated rather than stored

xspeedcruiser committed:
Changes for java chaincode implementation in chaincode_support.go

xspeedcruiser committed:
Changes for java chaincode implementation in chaincode_support.go

xspeedcruiser committed:
Changes for java chaincode implementation in chaincode_support.go

spomytkin committed:
sample java chaincode

spomytkin committed:
sample java chaincode

spomytkin committed:
sample java chaincode

xspeedcruiser committed:
Changes for java chaincode implementation in chaincode_support.go

xspeedcruiser committed:
Changes for java chaincode implementation in chaincode_support.go

xspeedcruiser committed:
Changes for java chaincode implementation in chaincode_support.go

spomytkin committed:
Sample logging with  org.apache.commons.logging

spomytkin committed:
Sample logging with  org.apache.commons.logging

spomytkin committed:
Sample logging with  org.apache.commons.logging

spomytkin committed:
Refactoring of shim package to commons-logging

spomytkin committed:
Refactoring of shim package to commons-logging

spomytkin committed:
Refactoring of shim package to commons-logging

xspeedcruiser committed:
Changes for deploying Java chaincode in non-dev or container mode

xspeedcruiser committed:
Changes for deploying Java chaincode in non-dev or container mode

xspeedcruiser committed:
Changes for deploying Java chaincode in non-dev or container mode

xspeedcruiser committed:
Fixed chaincode_support.go for CAR chaincode implementation

xspeedcruiser committed:
golint and govet changes

xspeedcruiser committed:
golint and govet changes

xspeedcruiser committed:
golint and govet changes

xspeedcruiser committed:
Added unit-test and behave tests for java-shim implementation

xspeedcruiser committed:
Added unit-test and behave tests for java-shim implementation

xspeedcruiser committed:
Merge branch 'javashim' of github.com:xspeedcruiser/fabric into javashim

xspeedcruiser committed:
Merged in chaincode_support.go with remote

xspeedcruiser committed:
Merged in chaincode_support.go with remote

Commit:
Fix memory leaks in busywork/bin/fabricLogger

binhn committed:
Merge pull request #1878 from jyellick/remove-obc-classic

Remove the unneeded obc-classic code

Commit:
Busywork: Changes to support latest blockchain semantics

The definition of correctness for **busywork** tests has been relaxed due to
the realities of the current PBFT consensus algorithms. Due to state transfer,
we can no longer assume that each peer has actually executed each
transaction. Instead, we can only demand that the state of the peer is
consistent with having executed every transaction. We also now wait until the
peers reach a consensus on the blockchain height at the end of the run before
querying for final status, since queries now fail on peers that are in the
middle of state transfer. Once a couple of outstanding consensus PRs are
processed we will probably be able to state that the busywork/counters test
case *stress2b* is working.

Commit:
A few more updates for busy621

xspeedcruiser committed:
Peer address fix sent to docker env during chaincode creation

xspeedcruiser committed:
Peer address fix sent to docker env during chaincode creation

srderson committed:
Merge pull request #1944 from bcbrock/fabricLogger-leaks

Fix memory leaks in busywork/bin/fabricLogger

srderson committed:
Merge pull request #1946 from bcbrock/busy621

Busywork: Changes to allow PBFT batch runs to pass busywork tests

xspeedcruiser committed:
Removed gradle compile before run when starting the chaincode
Modified default chaincode startup timeout for gradle to download dependencies and build
Corrected Behave tests for java shim example02

xspeedcruiser committed:
Removed gradle compile before run when starting the chaincode
Modified default chaincode startup timeout for gradle to download dependencies and build
Corrected Behave tests for java shim example02

xspeedcruiser committed:
Moved java-shim to core/chaincode/shim

xspeedcruiser committed:
Moved java-shim to core/chaincode/shim

xspeedcruiser committed:
Modified unit test to reflect change in java shim structure

xspeedcruiser committed:
Modified unit test to reflect change in java shim structure

xspeedcruiser committed:
Removed vagrant check in unit test

xspeedcruiser committed:
Removed vagrant check in unit test

xspeedcruiser committed:
Moved java shim behave test to the existing file

xspeedcruiser committed:
Moved java shim behave test to the existing file

xspeedcruiser committed:
Re-organized java packages

xspeedcruiser committed:
Re-organized java packages

xspeedcruiser committed:
Merge branch 'javashim' of github.com:xspeedcruiser/fabric into javashim

xspeedcruiser committed:
Merge branch 'javashim' of github.com:xspeedcruiser/fabric into javashim

xspeedcruiser committed:
Modified example to reflect new structure, and included basic documentation

xspeedcruiser committed:
Update JAVAChaincode.md

xspeedcruiser commented:
Signed-off-by: satheesh.ceg@gmail.com

xspeedcruiser committed:
goimport errors

xspeedcruiser committed:
Merge branch 'javashim' of github.com:xspeedcruiser/fabric into javashim

xspeedcruiser committed:
Merge remote-tracking branch 'upstream/master' into javashim

xspeedcruiser committed:
Merge remote-tracking branch 'upstream/master' into javashim

xspeedcruiser committed:
Modified example paths to relative in unit test and behave test

xspeedcruiser committed:
Merge remote-tracking branch 'upstream/master' into javashim

mastersingh24 commented:
I'll try to actually test this out this weekend.  I'd like to make sure we get several people try to actually build and deploy Java chaincode before pulling this in but overall this is going to be a big boost for the project.

srderson commented:
One quick note, all source files need to include the license in a comment at the top of the file.

```
/*
Copyright <your company> 2016 All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

         http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
```

I was surprised that I didn't see instructions with this info in https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md. @christo4ferris Are these instructions somewhere else or do we need to add them?

christo4ferris commented:
@srderson it is actually in the Coding Golang section... of course, it needs to be broader than that as it applies to all source files. I'll send a PR to update the CONTRIBUTING.md to make that all clearer.

muralisrini commented on core/chaincode/chaincode_support.go:
```diff
@@ -289,17 +290,30 @@ func (chaincodeSupport *ChaincodeSupport) getArgsAndEnv(cID *pb.ChaincodeID) (ar
+	case pb.ChaincodeSpec_CAR:
+		//chaincode executable will be same as the name of the chaincode
+		args = []string{chaincodeSupport.chaincodeInstallPath + cID.Name, fmt.Sprintf("-peer.address=%s", chaincodeSupport.peerAddress)}
+		chaincodeLogger.Debugf("Executable is %s", args[0])
```
a nit, but perhaps change the case to something like this ?

```

case pb.ChaincodeSpec_GOLANG:
     fallthrough
case pb.ChaincodeSpec_CAR:
           //chaincode executable will be same as the name of the chaincode
        args = []string{chaincodeSupport.chaincodeInstallPath + cID.Name, fmt.Sprintf("-peer.address=%s", chaincodeSupport.peerAddress)}
        chaincodeLogger.Debugf("Executable is %s", args[0])
case pb.ChaincodeSpec_JAVA:
    ...

```

xspeedcruiser replied:
Yes, I will address this and update the PR.

muralisrini commented:
@xspeedcruiser thanks, will review this over next few days

muralisrini commented on scripts/provision/docker.sh:
```diff
@@ -98,6 +98,14 @@ done`
+RUN ln -s /opt/gradle-2.12/bin/gradle /usr/bin
+# TODO JAVA_HOME set here, consider using update-java-alternatives
+ENV JAVA_HOME /usr/lib/jvm/java-1.8.0-openjdk-amd64
+
```
@xspeedcruiser Doesn't have to be done immediately, but if/when we create an image for java we should move this out of baseimage

xspeedcruiser replied:
@muralisrini .. Sure, This will be taken care when creating a separate image for Java lang base image. I am planning to add the process of downloading dependent libraries  and building java shim layer libraries during docker image building to reduce the chaincode deployment time, and chaincode deployment can refer to those libraries inside the docker image for build and runtime.

muralisrini commented on docs/dev-setup/JAVAChaincode.md:
```diff
@@ -0,0 +1,71 @@
+
+	cd $GOPATH/src/github.com/hyperledger/fabric/core/chaincode/shim/java
+
+ and run 'gradle build'
```
@xspeedcruiser step 5) is not necessary for running the java chaincode in docker (ie, not in "--peer-chaincodedev" mode). 

In fact, we need to add instructions to run in "--peer-chaincodedev" mode. To do that we will have to install gradle and java 1.8 in vagrant.

xspeedcruiser replied:
@muralisrini I will add the instructions for running it in DEV mode

muralisrini commented:
@xspeedcruiser The following worked off the bat under vagrant ! Very nice.

```
cd $GOPATH/src/github.com/hyperledger/fabric/peer

$ ./peer chaincode deploy -l java -p /opt/gopath/src/github.com/hyperledger/fabric/core/chaincode/shim/java -c '{"Function": "init", "Args": ["a","100", "b", "200"]}'
2016/06/27 01:06:15 Load docker HostConfig: &{Binds:[] CapAdd:[] CapDrop:[] ContainerIDFile: LxcConf:[] Privileged:false PortBindings:map[] Links:[] PublishAllPorts:false DNS:[] DNSSearch:[] ExtraHosts:[] VolumesFrom:[] NetworkMode:host IpcMode: PidMode: UTSMode: RestartPolicy:{Name: MaximumRetryCount:0} Devices:[] LogConfig:{Type: Config:map[]} ReadonlyRootfs:false SecurityOpt:[] CgroupParent: Memory:0 MemorySwap:0 MemorySwappiness:0 OOMKillDisable:false CPUShares:0 CPUSet: CPUSetCPUs: CPUSetMEMs: CPUQuota:0 CPUPeriod:0 BlkioWeight:0 Ulimits:[]}
01:06:15.907 [crypto] main -> INFO 002 Log level recognized 'info', set to INFO
52edc363618bdedbbdc7ffe51044f103d58e04fb55f708d14e4b435b64df1187b60225c3f2d4cf4e78bf72ba896da9e19f9e8c1ac1c8a42f44c51d18f91ffa07

$ ./peer chaincode invoke -l java -n 52edc363618bdedbbdc7ffe51044f103d58e04fb55f708d14e4b435b64df1187b60225c3f2d4cf4e78bf72ba896da9e19f9e8c1ac1c8a42f44c51d18f91ffa07 -c '{"Function": "transfer", "Args": ["a","b", "20"]}'
2016/06/27 01:07:16 Load docker HostConfig: &{Binds:[] CapAdd:[] CapDrop:[] ContainerIDFile: LxcConf:[] Privileged:false PortBindings:map[] Links:[] PublishAllPorts:false DNS:[] DNSSearch:[] ExtraHosts:[] VolumesFrom:[] NetworkMode:host IpcMode: PidMode: UTSMode: RestartPolicy:{Name: MaximumRetryCount:0} Devices:[] LogConfig:{Type: Config:map[]} ReadonlyRootfs:false SecurityOpt:[] CgroupParent: Memory:0 MemorySwap:0 MemorySwappiness:0 OOMKillDisable:false CPUShares:0 CPUSet: CPUSetCPUs: CPUSetMEMs: CPUQuota:0 CPUPeriod:0 BlkioWeight:0 Ulimits:[]}
01:07:16.499 [crypto] main -> INFO 002 Log level recognized 'info', set to INFO
1c2fcce5-cc95-4512-87fb-87464679c3ab

$ ./peer chaincode query -l java -n 52edc363618bdedbbdc7ffe51044f103d58e04fb55f708d14e4b435b64df1187b60225c3f2d4cf4e78bf72ba896da9e19f9e8c1ac1c8a42f44c51d18f91ffa07 -c '{"Function": "query", "Args": ["a"]}'
2016/06/27 01:07:23 Load docker HostConfig: &{Binds:[] CapAdd:[] CapDrop:[] ContainerIDFile: LxcConf:[] Privileged:false PortBindings:map[] Links:[] PublishAllPorts:false DNS:[] DNSSearch:[] ExtraHosts:[] VolumesFrom:[] NetworkMode:host IpcMode: PidMode: UTSMode: RestartPolicy:{Name: MaximumRetryCount:0} Devices:[] LogConfig:{Type: Config:map[]} ReadonlyRootfs:false SecurityOpt:[] CgroupParent: Memory:0 MemorySwap:0 MemorySwappiness:0 OOMKillDisable:false CPUShares:0 CPUSet: CPUSetCPUs: CPUSetMEMs: CPUQuota:0 CPUPeriod:0 BlkioWeight:0 Ulimits:[]}
01:07:23.292 [crypto] main -> INFO 002 Log level recognized 'info', set to INFO
{"Name":"a","Amount":"80"}

```

I have added a few inline comments (there maybe some more).

Another important organization we probably will have to do -

The chaincode examples (each with its own "build.gradle") should be under fabric/examples/chaincode/java/ with none of the common shim code in that folder. "build.gradle" in each java chaincode would refer to the common shim code in the fabric. Thus the build.gradle in the examples will become a template for java chaincode development.

xspeedcruiser commented:
@srderson I will add the copyright notice to all the new files.
@muralisrini I have plans to decouple the java shim layer library and chaincode building process. The shim layer libraries can be built when the java lang base docker image is created. Then have another gradle build file for the chaincodes, that will depend on the java shim library and build it to an executable jar. This will help us with 2 things.
1. Eliminate grade dependency inside the docker image, making it slightly thinner
2. The entry class name for executing the java chaincode need not be passed from peer when deploying the chaincode. We are currently sending the root folder path to the chaincode package from the peer in deploy transaction.

xspeedcruiser committed:
Changed the case statement to have a single path for CAR and GOLANG

xspeedcruiser committed:
Added license text

xspeedcruiser committed:
Updated java chaincode documentation for running it under dev mode

xspeedcruiser committed:
Merge remote-tracking branch 'upstream/master' into javashim

xspeedcruiser committed:
Updated chaincode doc for better readability

xspeedcruiser committed:
Updated behave test scenario name and removed comments

xspeedcruiser committed:
Modified host.sh for Vagrant box gradle and jdk setup

xspeedcruiser committed:
Gradle sym link creation inside Vagrant env

xspeedcruiser committed:
Merge remote-tracking branch 'upstream/master' into javashim

xspeedcruiser committed:
Fixed type in SimpleSample

xspeedcruiser committed:
KEEPALIVE changes in the java shim layer

TODO: chaincode.proto is checked into repo, remove it and copy the file as part of build process

xspeedcruiser committed:
temporary commit for chaincodeevent.proto

xspeedcruiser commented:
@srderson @muralisrini KEEPALIVE changes also checked in, Please review and comment.

muralisrini commented:
The --peer-chaincodedev instructions worked.

This is a fairly significant PR in size and function and can be divided into the following parts
- changes to mainline fabric code (mainly in chaincode component)
- addition of shim/java component
- new behave function
- documentation

The changes to the mainline fabric code are not much and do not interfere with non-java paths.
We know there's some planned work in refactorization to the shim/java component in future PRs.
There's probably going to be cleanup/enhancments to the documentation but the instructions for samples work.

Based on the above, recommend this be merged.

spomytkin committed:
"fine" and "debug"  mesages in base and handler, conf. file stab

srderson commented:
@xspeedcruiser Could you add some basic documentation? Possibly we need a /docs/chaincode folder with a `Java Chaincode.md` file under it? This could describe basic of deploying a Java chaincode.

srderson commented:
@xspeedcruiser Never mind, I completely missed `JAVAChaincode.md` on my first review.

srderson commented:
Confirmed things pass locally. A `vagrant destroy` `vagrant up` is not required for those who are not running Java chaincode. LGTM

@spomytkin @xspeedcruiser - could you please open issues in GitHub for the other items that need to be completed? I will add a 'java chaincode' tag to these
